projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/c/LLVMSignal.java

```diff
@@ -0,0 +1,425 @@
+                isRunning.set(true);
+                context.registerThread(this);
+            } catch (IllegalArgumentException e) {
+                stack.free();
+                throw e;
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        public boolean isRunning() {
+            return isRunning.get();
+        }
+
+        @Override
+        @TruffleBoundary
+        protected void finalize() throws Throwable {
+            super.finalize();
+            stop();
+            unregisterFromContext();
+        }
+
+        private static final long HANDLE_MAX_WAITING_TIME = 250; // ms
+
+        @Override
+        public void handle(Signal arg0) {
+            try {
+                if (!globalSignalHandlerLock.tryLock(HANDLE_MAX_WAITING_TIME, TimeUnit.MILLISECONDS)) {
+                    LLVMLogger.error("could not execute signal handler. Sulong can currently only execute one signal at once!");
+                    return;
+                }
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+            lock.lock();
+            try {
+                if (isRunning.get()) {
+                    callTarget.call();
+                }
+            } finally {
+                lock.unlock();
+                globalSignalHandlerLock.unlock();
+            }
+
+            // probably, this was our last turn for the signal handler
+            if (!isRunning.get()) {
+                unregisterFromContext();
+            }
+        }
+
+        public LLVMFunctionDescriptor getFunction() {
+            return function;
+        }
+
+        /**
+         * Required to call if a new LLVMSignalHandler take over the signal. Otherwise it would
+         * unregister the signal when this Object is going to be deallocated or stopped.
+         */
+        private void setStopped() {
+            isRunning.set(false);
+
+            /*
+             * Either no handler is currently running, or it would unregister after finishing it's
+             * execution.
+             */
+            tryUnregisterFromContext();
+        }
+
+        @Override
+        public void stop() {
+            if (isRunning.getAndSet(false)) {
+                /*
+                 * it seems like we don't want to catch this signal anymore, as well as there is no
+                 * other signal handler which want to catch it. So we simply reset it to look like
+                 * no signal handler was registered at all.
+                 */
+                Signal.handle(signal, SignalHandler.SIG_DFL);
+            }
+
+            /*
+             * Either no handler is currently running, or it would unregister after finishing it's
+             * execution.
+             */
+            tryUnregisterFromContext();
+        }
+
+        @Override
+        public void awaitFinish() {
+            stop();
+
+            // wait until handle is finished
+            lock.lock();
```

Comment:
yes, as I had written in the comment above, it waits until no lock is held by the object. Means either some little configuration step finished, or more important, no signal handler is executed anymore.
Because we set ```isRunning``` to false before, execution of the handler would exit without calling the referenced signal handler, which means effectivly, the handler is now deactivated and we can safely unregister it.